File: src/platform_context_graph/data_intelligence/dbt_sql_identifiers.py
"""Identifier-scanning helpers for compiled dbt SQL lineage."""

from __future__ import annotations

import re

_IDENTIFIER_RE = re.compile(r"\b(?P<identifier>[A-Za-z_][A-Za-z0-9_]*)\b")
_SINGLE_QUOTED_STRING_RE = re.compile(r"'(?:[^'\\\\]|\\\\.)*'")
_IGNORED_UNQUALIFIED_IDENTIFIERS = {
    "and",
    "as",
    "asc",
    "case",
    "cast",
    "coalesce",
    "count",
    "desc",
    "distinct",
    "else",
    "end",
    "false",
    "from",
    "group",
    "in",
    "is",
    "join",
    "left",
    "like",
    "limit",
    "lower",
    "not",
    "null",
    "on",
    "or",
    "order",
    "over",
    "partition",
    "by",
    "rows",
    "range",
    "preceding",
    "following",
    "current",
    "row",
    "right",
    "select",
    "sum",
    "then",
    "true",
    "upper",
    "when",
    "where",
    "with",
}


def unqualified_identifiers(
    expression: str,
    *,
    matched_identifiers: set[str],
) -> list[str]:
    """Return bare identifier candidates that still need lineage resolution."""

    sanitized_expression = _SINGLE_QUOTED_STRING_RE.sub(
        lambda match: " " * len(match.group(0)),
        expression,
    )
    identifiers: list[str] = []
    seen_identifiers: set[str] = set()
    for match in _IDENTIFIER_RE.finditer(sanitized_expression):
        identifier = match.group("identifier")
        lowered = identifier.lower()
        if (
            identifier in matched_identifiers
            or lowered in _IGNORED_UNQUALIFIED_IDENTIFIERS
            or identifier in seen_identifiers
        ):
            continue
        next_non_space = _next_non_space_character(sanitized_expression, match.end())
        if next_non_space == "(":
            continue
        seen_identifiers.add(identifier)
        identifiers.append(identifier)
    return identifiers


def _next_non_space_character(text: str, index: int) -> str | None:
    """Return the next non-whitespace character after one expression offset."""

    while index < len(text):
        if not text[index].isspace():
            return text[index]
        index += 1
    return None


__all__ = ["unqualified_identifiers"]

File: go/internal/parser/dbt_sql_identifiers.go
package parser

import (
	"regexp"
	"strings"
)

var (
	dbtIdentifierRe         = regexp.MustCompile(`\b(?P<identifier>[A-Za-z_][A-Za-z0-9_]*)\b`)
	dbtSingleQuotedStringRe = regexp.MustCompile(`'(?:[^'\\]|\\.)*'`)
	dbtIgnoredIdentifiers   = map[string]struct{}{
		"and": {}, "as": {}, "asc": {}, "case": {}, "cast": {}, "coalesce": {}, "count": {}, "desc": {}, "distinct": {}, "else": {},
		"end": {}, "false": {}, "from": {}, "group": {}, "in": {}, "is": {}, "join": {}, "left": {}, "like": {}, "limit": {},
		"lower": {}, "not": {}, "null": {}, "on": {}, "or": {}, "order": {}, "over": {}, "partition": {}, "by": {}, "rows": {},
		"range": {}, "preceding": {}, "following": {}, "current": {}, "row": {}, "right": {}, "select": {}, "sum": {}, "then": {},
		"true": {}, "upper": {}, "when": {}, "where": {}, "with": {},
	}
)

func unqualifiedIdentifiers(expression string, matchedIdentifiers map[string]struct{}) []string {
	sanitized := dbtSingleQuotedStringRe.ReplaceAllStringFunc(expression, func(value string) string {
		return regexp.MustCompile(`.`).ReplaceAllString(value, " ")
	})
	identifiers := make([]string, 0)
	seen := make(map[string]struct{})
	for _, match := range dbtIdentifierRe.FindAllStringSubmatchIndex(sanitized, -1) {
		identifier := sanitized[match[2]:match[3]]
		lowered := strings.ToLower(identifier)
		if _, ok := matchedIdentifiers[identifier]; ok {
			continue
		}
		if _, ok := dbtIgnoredIdentifiers[lowered]; ok {
			continue
		}
		if _, ok := seen[identifier]; ok {
			continue
		}
		if next := nextNonSpaceCharacter(sanitized, match[1]); next == "(" || next == "." {
			continue
		}
		seen[identifier] = struct{}{}
		identifiers = append(identifiers, identifier)
	}
	return identifiers
}

func nextNonSpaceCharacter(text string, index int) string {
	for index < len(text) {
		if text[index] != ' ' && text[index] != '\t' && text[index] != '\n' && text[index] != '\r' {
			return string(text[index])
		}
		index++
	}
	return ""
}
